File: Lecture/Binary Trees/PrintLevelWise.cpp
/***
 * For a given a Binary Tree of type integer, print the complete information of every node, when traversed in a level-order fashion.
 * To print the information of a node with data D, you need to follow the exact format :

           D:L:X,R:Y

    Where D is the data of a node present in the binary tree. 
    X and Y are the values of the left(L) and right(R) child of the node.
    Print -1 if the child doesn't exist.
 ***/
/**********************************************************
	Following is the Binary Tree Node class structure
	template <typename T>
	class BinaryTreeNode {
    	public : 
    	T data;
    	BinaryTreeNode<T> *left;
    	BinaryTreeNode<T> *right;

    	BinaryTreeNode(T data) {
        	this -> data = data;
        	left = NULL;
        	right = NULL;
    	}
	};
***********************************************************/
void printLevelWise(BinaryTreeNode<int> *root) {
	if(root == NULL)
        return;
    
    queue<BinaryTreeNode<int>*> pendingNodes;
    pendingNodes.push(root);
    
    while(pendingNodes.size() != 0) {
        BinaryTreeNode<int>* front = pendingNodes.front();
        pendingNodes.pop();
        cout << front->data << ":L:";
        
        if(front->left) {
            cout << front->left->data;
            pendingNodes.push(front->left);
        }
        else
            cout << "-1";
        
        cout << ",R:";
        
        if(front->right) {
            cout << front->right->data;
            pendingNodes.push(front->right);
        }
        else
            cout << "-1";
        cout << '\n';
    }
}

File: Lecture/Binary Trees/PrintLevelWise.java
/*
Time complexity: O(N)
Space complexity: O(N)
where N is the number of nodes in the input tree
*/
import java.util.Queue;
import java.util.LinkedList;

public class PrintLevelWise {

    class BinaryTreeNode<T> {
        T data;
        BinaryTreeNode<T> left;
        BinaryTreeNode<T> right;
    
        public BinaryTreeNode(T data) {
            this.data = data;
            this.left = null;
            this.right = null;
        }
    }
    
    public static void printLevelWise(BinaryTreeNode<Integer> root) {
        Queue<BinaryTreeNode<Integer>> pendingNodes = new LinkedList<>();
        pendingNodes.add(root);
        while (!pendingNodes.isEmpty()) {
            BinaryTreeNode<Integer> frontNode = pendingNodes.poll();
            if (frontNode == null) {
                System.out.println();
                if (!pendingNodes.isEmpty()) {
                    pendingNodes.add(null);
                }
            } else {
                System.out.print(frontNode.data + ":L:");
                if (frontNode.left != null) {
                    pendingNodes.add(frontNode.left);
                    System.out.print(frontNode.left.data + ",R:");
                } else {
                    System.out.print(-1 + ",R:");
                }
                if (frontNode.right != null) {
                    pendingNodes.add(frontNode.right);
                    System.out.println(frontNode.right.data);
                } else {
                    System.out.println(-1);
                }
            }
        }
    }
}